exec hangs when program not found

Michael,

with the exec package, it seems to hang when the program to run is not found. 

Debugger shows that the IOexception was not caught and things kept moving forward and the code gets stuck at a wait() in ExecuteWatchdog:ensureStarted(), line 114. So the ManagedProcessExeception was never thrown.

Can you look at [1] to see if we are calling this correctly.

Thanks, Sam

[1] https://github.com/shague/opendaylight-ansible/blob/master/southbound/impl/src/main/java/org/opendaylight/ansible/southbound/AnsibleCommandServiceImpl.java#L82

[main] INFO ch.vorburger.exec.ManagedProcess - Starting Program [ansible-runner, -j, --hosts, localhost, -p, file, run, path] (in working directory /home/shague/git/ansible/southbound/impl/.)
[Exec Default Executor] ERROR ch.vorburger.exec.LoggingExecuteResultHandler - Program [ansible-runner, -j, --hosts, localhost, -p, file, run, path] (in working directory /home/shague/git/ansible/southbound/impl/.) failed unexpectedly
org.apache.commons.exec.ExecuteException: Execution failed (Exit value: -559038737. Caused by java.io.IOException: Cannot run program "ansible-runner" (in directory "."): error=2, No such file or directory)
	at org.apache.commons.exec.DefaultExecutor$1.run(DefaultExecutor.java:205)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.IOException: Cannot run program "ansible-runner" (in directory "."): error=2, No such file or directory
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
	at java.lang.Runtime.exec(Runtime.java:620)
	at org.apache.commons.exec.launcher.Java13CommandLauncher.exec(Java13CommandLauncher.java:61)
	at org.apache.commons.exec.DefaultExecutor.launch(DefaultExecutor.java:279)
	at org.apache.commons.exec.DefaultExecutor.executeInternal(DefaultExecutor.java:336)
	at org.apache.commons.exec.DefaultExecutor.access$200(DefaultExecutor.java:48)
	at org.apache.commons.exec.DefaultExecutor$1.run(DefaultExecutor.java:200)
	... 1 more
Caused by: java.io.IOException: error=2, No such file or directory
	at java.lang.UNIXProcess.forkAndExec(Native Method)
	at java.lang.UNIXProcess.<init>(UNIXProcess.java:247)
	at java.lang.ProcessImpl.start(ProcessImpl.java:134)
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1029)
	... 7 more

--- a/pom.xml
+++ b/pom.xml
@@ -253,7 +253,7 @@
     <dependency>
       <groupId>org.apache.commons</groupId>
       <artifactId>commons-exec</artifactId>
-      <version>1.3</version>
+      <version>1.4.0</version>
     </dependency>
     <dependency>
       <!-- Just for its CircularFifoQueue... -->

--- a/src/main/java/ch/vorburger/exec/ExtendedDefaultExecutor.java
+++ b/src/main/java/ch/vorburger/exec/ExtendedDefaultExecutor.java
@@ -22,6 +22,7 @@ package ch.vorburger.exec;
 import java.util.function.Function;
 import org.apache.commons.exec.DefaultExecutor;
 
+@SuppressWarnings("deprecation") // TODO https://github.com/vorburger/ch.vorburger.exec/issues/189
 class ExtendedDefaultExecutor extends DefaultExecutor {
 
     private Function<Integer, Boolean> exitValueChecker;

--- a/src/main/java/ch/vorburger/exec/ManagedProcess.java
+++ b/src/main/java/ch/vorburger/exec/ManagedProcess.java
@@ -108,6 +108,7 @@ public class ManagedProcess implements ManagedProcessState {
      * @param stdErrs StandardError from the console
      * @param listener A <tt>ManagedProcessListener</tt> which is notified when process completes or fails
      */
+    @SuppressWarnings("deprecation") // TODO https://github.com/vorburger/ch.vorburger.exec/issues/189
     ManagedProcess(CommandLine commandLine, File directory, Map<String, String> environment,
             InputStream input, boolean destroyOnShutdown, int consoleBufferMaxLines,
             OutputStreamLogDispatcher outputStreamLogDispatcher,

--- a/src/main/java/ch/vorburger/exec/ManagedProcessBuilder.java
+++ b/src/main/java/ch/vorburger/exec/ManagedProcessBuilder.java
@@ -88,7 +88,8 @@ public class ManagedProcessBuilder {
         environment = initialEnvironment();
     }
 
-    protected Map<String, String> initialEnvironment() throws ManagedProcessException {
+    // It's static due to: "possible 'this' escape before subclass is fully initialized"
+    protected static Map<String, String> initialEnvironment() throws ManagedProcessException {
         try {
             return EnvironmentUtils.getProcEnvironment();
         } catch (IOException e) {

--- a/src/main/java/ch/vorburger/exec/StopCheckExecuteWatchdog.java
+++ b/src/main/java/ch/vorburger/exec/StopCheckExecuteWatchdog.java
@@ -31,6 +31,7 @@ public class StopCheckExecuteWatchdog extends ExecuteWatchdog {
      *         the timeout for the process in milliseconds. It must be
      *         greater than 0 or 'INFINITE_TIMEOUT'
      */
+    @SuppressWarnings("deprecation") // TODO https://github.com/vorburger/ch.vorburger.exec/issues/189
     public StopCheckExecuteWatchdog(long timeout) {
         super(timeout);
     }

